File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/individuals/job_stage_controller.py
import logging
from django.contrib import messages
from django.db import transaction
from django.http import HttpRequest, HttpResponse
from django.shortcuts import get_object_or_404, redirect, render
from .._helpers.http import get_redirect_url
from .._traits.controller import Controller
from ....Models.individuals.job_stage import JobStage

logger = logging.getLogger(__name__)

class JobStageController(Controller):
  def index(self, request: HttpRequest) -> HttpResponse:
    try:
      self.authorize('viewAny', ['JobStage'])
      stages = JobStage.objects.filter(created_by=request.user.creator_id()).order_by('order')
      return render(request, 'job_stage/index.html', {'stages': stages})
    except Exception as e:
      logger.error(f"Method index failed to list job stages: {e}")
      messages.error(request, str(e))
      return redirect(get_redirect_url(request))

  def create(self, request: HttpRequest) -> HttpResponse:
    try:
      self.authorize('create', ['JobStage'])
      return render(request, 'job_stage/create.html')
    except Exception as e:
      logger.error(f"Method create failed to render form: {e}")
      messages.error(request, str(e))
      return redirect(get_redirect_url(request))

  def store(self, request: HttpRequest) -> HttpResponse:
    try:
      self.authorize('create', ['JobStage'])
      data = request.POST
      if not data.get('title'):
        messages.error(request, 'Title is required.')
        return redirect(get_redirect_url(request))
      with transaction.atomic():
        js = JobStage(title=data['title'], created_by=request.user.creator_id())
        js.save()
      messages.success(request, 'Job stage successfully created.')
      return redirect(request.META.get('HTTP_REFERER', get_redirect_url(request)))
    except Exception as e:
      logger.error(f"Method store failed to save job stage: {e}")
      messages.error(request, str(e))
      return redirect(get_redirect_url(request))

  def show(self, request: HttpRequest, job_stage_id: int) -> HttpResponse:
    from ....Models.individuals.job_application import JobApplication
    try:
      self.authorize('view', ['JobStage', job_stage_id])
      js = get_object_or_404(JobStage, pk=job_stage_id)
      applications = JobApplication.objects.filter(created_by=request.user.creator_id(), job_stage=js)
      return render(request, 'job_stage/show.html', {'job_stage': js, 'applications': applications})
    except Exception as e:
      logger.error(f"Method show failed: {e}")
      messages.error(request, str(e))
      return redirect(get_redirect_url(request))

  def edit(self, request: HttpRequest, job_stage_id: int) -> HttpResponse:
    try:
      self.authorize('update', ['JobStage', job_stage_id])
      js = get_object_or_404(JobStage, pk=job_stage_id)
      return render(request, 'job_stage/edit.html', {'job_stage': js})
    except Exception as e:
      logger.error(f"Method edit failed: {e}")
      messages.error(request, str(e))
      return redirect(get_redirect_url(request))

  def update(self, request: HttpRequest, job_stage_id: int) -> HttpResponse:
    try:
      self.authorize('update', ['JobStage', job_stage_id])
      data = request.POST
      if not data.get('title'):
        messages.error(request, 'Title is required.')
        return redirect(get_redirect_url(request))
      js = get_object_or_404(JobStage, pk=job_stage_id)
      js.title = data['title']
      js.created_by = request.user.creator_id()
      js.save()
      messages.success(request, 'Job stage successfully updated.')
      return redirect(request.META.get('HTTP_REFERER', get_redirect_url(request)))
    except Exception as e:
      logger.error(f"Method update failed: {e}")
      messages.error(request, str(e))
      return redirect(get_redirect_url(request))

  def destroy(self, request: HttpRequest, job_stage_id: int) -> HttpResponse:
    try:
      self.authorize('delete', ['JobStage', job_stage_id])
      js = get_object_or_404(JobStage, pk=job_stage_id)
      if js.created_by != request.user.creator_id():
        messages.error(request, 'Permission denied.')
        return redirect(get_redirect_url(request))
      js.delete()
      messages.success(request, 'Job stage successfully deleted.')
      return redirect(request.META.get('HTTP_REFERER', get_redirect_url(request)))
    except Exception as e:
      logger.error(f"Method destroy failed: {e}")
      messages.error(request, str(e))
      return redirect(get_redirect_url(request))

  def order(self, request: HttpRequest) -> HttpResponse:
    try:
      self.authorize('update', ['JobStage'])
      order_list = request.POST.getlist('order[]') or request.POST.getlist('order')
      for idx, item in enumerate(order_list):
        js = get_object_or_404(JobStage, pk=item)
        js.order = idx
        js.save()
      return HttpResponse(status=204)
    except Exception as e:
      logger.error(f"Method order failed to reorder job stages: {e}")
      messages.error(request, str(e))
      return redirect(get_redirect_url(request))

File: _inc/laravel/app/Http/Controllers/Individuals/JobStageController.php
<?php

namespace App\Http\Controllers\Individuals;

use App\Http\Controllers\Abstracts\Controller;

use App\Config\Constants\{
	DatabaseConstants as DC,
	UsersConstants as UC,
	ViewsConstants as VW
};
use App\Models\JobStage;
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{
	RedirectResponse,
	Request
};
use Illuminate\Support\Facades\{
	Log,
	Validator,
	View as ViewFacade
};
use Illuminate\View\View;
use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};

class JobStageController extends Controller
{

	use ChecksLogin, ChecksPermissions;
	private const REDIRECT_INDEX = '/';
	private const PERM_MANAGE = 'manage job stage';
	private const PERM_CREATE = 'create job stage';
	private const PERM_EDIT  = 'edit job stage';
	private const PERM_DELETE = 'delete job stage';
	public const IDX = 'index';
	public const CRT = 'create';
	public const STR = 'store';
	public const SHW = 'show';
	public const EDT = 'edit';
	public const UPD = 'update';
	public const DEL = 'destroy';


	public function index(Request $req): RedirectResponse|View
	{
		$action = __FUNCTION__;
		$method = __METHOD__;
		return $this->measureProfile($action, function () use ($req, $action, $method) {
			if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
			if (($c = self::guard($req, self::PERM_MANAGE, self::REDIRECT_INDEX)) !== true) return $c;
			try {
				$stages = JobStage::where(DC::COL_TABLE_CREATOR, $u->creatorId())->orderBy('order')->get();
				$view = VW::JB_STG . '.' . $action;
				if (!ViewFacade::exists($view)) return defaultUndefinedException($req, new \RuntimeException('View not found'), $method);
				Log::debug($method . ' loaded', ['count' => $stages->count()]);
				return ViewFacade::make($view, compact('stages'));
			} catch (\Throwable $e) {
				Log::error($method . ' failed', ['error' => $e->getMessage()]);
				return defaultUndefinedException($req, $e, $method);
			}
		}, []);
	}

	public function create(Request $req): RedirectResponse|View
	{
		$action = __FUNCTION__;
		$method = __METHOD__;
		return $this->measureProfile($action, function () use ($req, $action, $method) {
			if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
			if (($c = self::guard($req, self::PERM_CREATE, self::REDIRECT_INDEX)) !== true) return $c;
			$view = VW::JB_STG . '.' . $action;
			if (!ViewFacade::exists($view)) return defaultUndefinedException($req, new \RuntimeException('View not found'), $method);
			return ViewFacade::make($view);
		}, []);
	}

	public function show(Request $req, JobStage $jobStage): View|RedirectResponse
	{
		$action = __FUNCTION__;
		$cls = __CLASS__;
		$method = __METHOD__;
		return $this->measureProfile($action, function () use ($req, $jobStage, $action, $cls, $method) {
			Log::debug($method . ' start', [UC::COL_USER_ID => $req->user()->id, 'stage_id' => $jobStage->id]);
			try {
				if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
				if (($c = self::guard($req, self::PERM_MANAGE, self::REDIRECT_INDEX)) !== true) return $c;
				if ($jobStage->created_by !== $u->creatorId()) return defaultPermissionDenial($req, new AuthorizationException(), $method);
				$view = VW::JB_STG . '.' . $action;
				if (!ViewFacade::exists($view)) return defaultUndefinedException($req, new \RuntimeException('View not found'), $method);
				return ViewFacade::make($view, compact('jobStage'));
			} catch (\Throwable $e) {
				Log::error($method . ' failed', ['error' => $e->getMessage()]);
				return defaultUndefinedException($req, $e, $cls . '::' . $action);
			}
		}, ['stage_id' => $jobStage->id]);
	}

	public function store(Request $req): RedirectResponse
	{
		$action = __FUNCTION__;
		$method = __METHOD__;
		return $this->measureProfile($action, function () use ($req, $method) {
			if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
			if (($c = self::guard($req, self::PERM_CREATE, self::REDIRECT_INDEX)) !== true) return $c;
			$v = Validator::make($req->all(), ['title' => 'required']);
			if ($v->fails()) return redirect()->back()->with('error', $v->errors()->first());
			try {
				JobStage::create(['title' => $req->input('title'), DC::COL_TABLE_CREATOR => $u->creatorId()]);
				Log::debug($method . ' created');
				return redirect()->back()->with('success', __('Job stage successfully created.'));
			} catch (\Throwable $e) {
				Log::error($method . ' failed', ['error' => $e->getMessage()]);
				return defaultUndefinedException($req, $e, $method);
			}
		}, ['title' => $req->input('title')]);
	}

	public function edit(Request $req, JobStage $jobStage): RedirectResponse|View
	{
		$action = __FUNCTION__;
		$method = __METHOD__;
		return $this->measureProfile($action, function () use ($req, $jobStage, $action, $method) {
			if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
			if (($c = self::guard($req, self::PERM_EDIT, self::REDIRECT_INDEX)) !== true) return $c;
			$view = VW::JB_STG . '.' . $action;
			if (!ViewFacade::exists($view)) return defaultUndefinedException($req, new \RuntimeException('View not found'), $method);
			return ViewFacade::make($view, compact('jobStage'));
		}, ['stage_id' => $jobStage->id]);
	}

	public function update(Request $req, JobStage $jobStage): RedirectResponse
	{
		$action = __FUNCTION__;
		$method = __METHOD__;
		return $this->measureProfile($action, function () use ($req, $jobStage, $method) {
			if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
			if (($c = self::guard($req, self::PERM_EDIT, self::REDIRECT_INDEX)) !== true) return $c;
			$v = Validator::make($req->all(), ['title' => 'required']);
			if ($v->fails()) return redirect()->back()->with('error', $v->errors()->first());
			try {
				$jobStage->update(['title' => $req->input('title'), DC::COL_TABLE_CREATOR => $u->creatorId()]);
				Log::debug($method . ' updated', ['stage_id' => $jobStage->id]);
				return redirect()->back()->with('success', __('Job stage successfully updated.'));
			} catch (\Throwable $e) {
				Log::error($method . ' failed', ['error' => $e->getMessage()]);
				return defaultUndefinedException($req, $e, $method);
			}
		}, ['stage_id' => $jobStage->id, 'title' => $req->input('title')]);
	}

	public function destroy(Request $req, JobStage $jobStage): RedirectResponse
	{
		$action = __FUNCTION__;
		$method = __METHOD__;
		return $this->measureProfile($action, function () use ($req, $jobStage, $method) {
			if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
			if (($c = self::guard($req, self::PERM_DELETE, self::REDIRECT_INDEX)) !== true) return $c;
			if ($jobStage->created_by !== $u->creatorId()) return defaultPermissionDenial($req, new AuthorizationException(), $method);
			try {
				$jobStage->delete();
				Log::debug($method . ' deleted', ['stage_id' => $jobStage->id]);
				return redirect()->back()->with('success', __('Job stage successfully deleted.'));
			} catch (\Throwable $e) {
				Log::error($method . ' failed', ['error' => $e->getMessage()]);
				return defaultUndefinedException($req, $e, $method);
			}
		}, ['stage_id' => $jobStage->id]);
	}

	public const ORD = 'order';
	public function order(Request $req): void
	{
		$action = __FUNCTION__;
		$method = __METHOD__;
		$this->measureProfile($action, function () use ($req, $method) {
			if ((self::_checkLogin()) instanceof RedirectResponse) return null;
			if ((self::guard($req, self::PERM_EDIT, self::REDIRECT_INDEX)) !== true) return null;
			$orderList = $req->input('order', []);
			foreach ($orderList as $position => $id) {
				try {
					JobStage::whereKey($id)->update(['order' => $position]);
					Log::debug($method . ' reordered', ['id' => $id, 'position' => $position]);
				} catch (\Throwable $e) {
					Log::error($method . ' failed for id', ['id' => $id, 'error' => $e->getMessage()]);
				}
			}
			return null;
		}, ['order' => $req->input('order', [])]);
	}
}
